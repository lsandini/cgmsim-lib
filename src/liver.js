const logger = require('pino')();
const sinusRun = require('./sinus.js');
module.exports = function(env) {
    const ISF = parseInt(env.ISF) / 18;
    const CR = parseInt(env.CR);
    logger.info('ISF:', ISF, 'CR:', CR);

    // the sinus and cosinus numbers vary around 1, from 0.5 to 1.5:
    // sin starts at 1.0 at midnight, is max at 6AM, is again 1 at 12 AM, and minimums at 0.5 a 6 PM
    // cosin starts at 1.5 at midnight, is 1 at 6AM, is minimus at 0.5 12 AM, and is 1 again at 6 PM

    const sinusdata = sinusRun();
    const sinus = sinusdata.sinus;
    const cosinus = sinusdata.cosinus;
    logger.info('sinus: ', sinus);
    logger.info('cosinus: ', cosinus);


    // let's simulate the carb impact of the liver, producing 10g of carbs / hour
    // if the ISF is 2 mmol/l/U,
    // and the CR is 10g/U,
    // => the the CF (carb factor) is 0.2 mmol/l/ 1g
    // so the BG increases 2 mmol/l/h, (every time 10g of carbs are delivered)

    // 0.2 mmol/l/h *10g /12 periods => bgi every 5 minutes or 0,166666 mmol/l/5min

    // by multiplying the liver_bgi by the sin function, the liver loog glucose production varies in a sinusoidal
    // form, being maximal at 6 AM and minimal ad 6 PM

    const liver = (ISF / CR) * (10 / 60);
    //const liver = 0.1666;
    const liver_sin = liver * sinus;
    logger.info('liver: ', liver);
    logger.info('liver_sin: ', liver_sin);



    // const fs = require('fs');
    // fs.writeFile('./files/latest_liver.json', liverString, function (err) {
    // 	if (err) logger.info('error', err);
    // });
    return liver_sin;
};