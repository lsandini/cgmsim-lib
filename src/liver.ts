import sinusRun from './sinus';
import logger from './utils';

export default function (
	isfConstant: number,
	cr: number,
	activities: { physical: number; alcohol: number },
	weight: number,
	timeZone: string,
): number {
	const _ISF = isfConstant;
	const _CR = cr;
	const activityFactor = activities?.physical >= 0 ? activities.physical : 1;
	const alcoholFactor = activities?.alcohol >= 0 ? activities.alcohol : 1;
	const { sinus } = sinusRun(timeZone);
	const CF = _ISF / _CR;
	const glucosePerMinute = 0.002 * weight;
	const liver = alcoholFactor * activityFactor * CF * glucosePerMinute; //(mmols/l)/min

	const liver_sin = liver * sinus;
	logger.debug('liver:  %o', liver);
	logger.debug('liver_sin:  %o', liver_sin);

	return liver_sin; //(mmol/l)/min
}
