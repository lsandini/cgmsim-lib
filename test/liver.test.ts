import liver from '../src/liver';
describe('Liver test', () => {
  beforeAll(() => {
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date('2001-01-01'));
  });

  afterAll(() => {
    jest.useRealTimers();
  });
  it.each([[[30 / 18, 10]], [[20 / 18, 8]], [[40 / 18, 14]]])(
    'test liver %p',
    ([isf, cr]) => {
      const r = liver(
        isf,
        cr,
        { physical: 1, alcohol: 0 },
        250 / 3,
        'Europe/Helsinki', //2001-01-01 UTC+2
      );
      expect(r).toMatchSnapshot();
    },
  );
});
