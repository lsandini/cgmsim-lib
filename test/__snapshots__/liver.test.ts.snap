// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Liver test test liver [1.1111111111111112, 8] 1`] = `0`;

exports[`Liver test test liver [1.6666666666666667, 10] 1`] = `0`;

exports[`Liver test test liver [2.2222222222222223, 14] 1`] = `0`;

exports[`Liver test test liver [20, 8] 1`] = `0.025462962962962965`;

exports[`Liver test test liver [30, 10] 1`] = `0.03055555555555556`;

exports[`Liver test test liver [40, 14] 1`] = `0.029100529100529106`;
